### PR TITLE
add initial knowledge doc and its script

### PR DESCRIPTION
**Pull request by TomasTomecek**

### CODE REVIEW

File: scripts/compile_ilab_qa.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+
+raw = []
+# ilab doesn't allow duplicate entries, so we need to make our entries unique
+haz_snippets = set()
+for file in glob.glob(f"{LOGDETECTIVE_DATA_DIR}/**/*.json", recursive=True):
+    with open(file) as f:
+        raw.append(json.load(f))
+
+
+def smart_string_split(s) -> Optional[str]:
```
Comment (jpodivin) [inline scripts/compile_ilab_qa.py]: I recommend using standard library textwrap https://docs.python.org/3/library/textwrap.html

File: scripts/compile_ilab_qa.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    # 112 = 7 spaces for padding, 112 the log line, 1 = EOL
+    if len(before) >= 112 or len(after) >= 112:
+        return f"{smart_string_split(before)}\n{smart_string_split(after)}"
+    else:
+        return f"{before}\n{after}"
+
+
+for e in raw:
+    for k, v in e['logs'].items():
+        for s in v['snippets']:
```
Comment (jpodivin) [inline scripts/compile_ilab_qa.py:52]: I would recommend to turn this into a set straight away. We can avoid all the checking that way, and no further add operations will be needed. Technically speaking, the O complexity should be the same, and the code cleaner.

I would also recommend filtering the `None` snippets before the main loop. That way the condition on None won't be necessary either.

**Review by jpodivin — CHANGES_REQUESTED** [review]

File: scripts/compile_ilab_qa.py (hunk 2 above)
Comment (TomasTomecek) [reply scripts/compile_ilab_qa.py:52]: I understand what you are asking me to do but have no idea how to achieve that.

`s` is dict and it's not hashable

File: scripts/compile_ilab_qa.py (hunk 1 above)
Comment (TomasTomecek) [reply scripts/compile_ilab_qa.py]: good suggestion, thanks, wasn't aware of this

File: scripts/compile_ilab_qa.py (hunk 2 above)
Comment (jpodivin) [reply scripts/compile_ilab_qa.py:52]: Right, I confused the `s` with `snippet`. Please disregard that.

Comment (TomasTomecek) [reply scripts/compile_ilab_qa.py:52]: I'll try to use single letter variables less :) I understand it's confusing for review; one of my bad habits

**Comment by TomasTomecek** [comment]
wow, the wrapping works so well, very good suggestion @jpodivin 

I'm not sure how relevant is whitespace for LLMs because the wrapping breaks some newlines - don't know how big of a concern this is

I added the change as a new commit so we can compare the wrapping

**Comment by jpodivin** [comment]
@TomasTomecek  There is a way to handle whitespace with `replace_whitespace` and `drop_whitespace` args. [1]
I'm not sure if it's going to achieve exactly the same result, but it could bring you closer. That being said, I don't think whitespace tokens have that much a of an importance for the model.

 
[1] https://docs.python.org/3/library/textwrap.html#textwrap.TextWrapper.replace_whitespace

File: scripts/compile_ilab_qa.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+            snippet = v['content'][s['start_index']:s['end_index']]
+            if len(snippet) > 150:
+                # too big, we'll figure it out later
+                continue
+            # 120 is the instructlab limit for a yaml line
+            # 112 = 7 spaces for padding, 112 the log line, 1 = EOL
+            # since snippet is the log chunk, we wanna be as strict as possible on the wrapping
+            snippet = "\n".join(textwrap.wrap(
```
Comment (jpodivin) [inline scripts/compile_ilab_qa.py]: fill[2] would be a bit cleaner. 

[2] https://docs.python.org/3/library/textwrap.html#textwrap.fill

File: scripts/compile_ilab_qa.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+}
+
+raw = []
+# ilab doesn't allow duplicate entries, so we need to make our entries unique
+haz_snippets = set()
+for file in glob.glob(f"{LOGDETECTIVE_DATA_DIR}/**/*.json", recursive=True):
+    with open(file) as f:
+        raw.append(json.load(f))
+
+for e in raw:
```
Comment (jpodivin) [inline scripts/compile_ilab_qa.py:50]: Given the many calls to the `wrap` function. I would recommend instantiating a TextWrapper object[3].
The performance gain may be marginal however.

[3] https://docs.python.org/3/library/textwrap.html#textwrap.TextWrapper

**Review by jpodivin — APPROVED** [review]
Minute nitpicks, but otherwise this seems good to go.

File: scripts/compile_ilab_qa.py (hunk 3 above)
Comment (TomasTomecek) [reply scripts/compile_ilab_qa.py]: neat, that's a nice shortcut

**Review by jpodivin — APPROVED** [review]

**Pull request merged**